src/OpenDebugAD7/TracepointManager.cs

```diff
@@ -0,0 +1,229 @@
+            }
+        }
+
+        internal void Replace(uint breakpointId, string newLogMesssage)
+        {
+            lock (_lock)
+            {
+                m_breakpointLogMessages[breakpointId] = newLogMesssage;
+            }
+        }
+
+        internal bool Contains(uint breakpointId)
+        {
+            lock (_lock)
+            {
+                return m_breakpointLogMessages.ContainsKey(breakpointId);
+            }
+        }
+
+        internal bool TryGetValue(uint breakpointId, out string logMessage)
+        {
+            lock (_lock)
+            {
+                return m_breakpointLogMessages.TryGetValue(breakpointId, out logMessage);
+            }
+        }
+
+        internal string GetLogMessage(uint breakpointId, IDebugThread2 pThread, uint radix)
+        {
+            string logMessage;
+            if (!this.TryGetValue(breakpointId, out logMessage))
+            {
+                return string.Empty;
+            }
+
+            // There is strings to interpolate in the log message.
+            if (regex.IsMatch(logMessage))
+            {
+                logMessage = GetInterpolatedLogMessage(logMessage, pThread, radix);
+            }
+
+            return logMessage;
+        }
+
+        private string GetInterpolatedLogMessage(string logMessage, IDebugThread2 pThread, uint radix)
+        {
+            if (pThread == null)
+            {
+                return logMessage;
+            }
+
+            // Get topFrame
+            IEnumDebugFrameInfo2 frameInfoEnum;
+            int hr = pThread.EnumFrameInfo(enum_FRAMEINFO_FLAGS.FIF_FRAME | enum_FRAMEINFO_FLAGS.FIF_FLAGS, Constants.EvaluationRadix, out frameInfoEnum);
+            if (hr < 0)
+            {
+                return logMessage;
+            }
+
+            FRAMEINFO[] topFrame = new FRAMEINFO[1];
+            uint fetched = 0;
+            hr = frameInfoEnum.Next(1, topFrame, ref fetched);
+            if (hr < 0 || fetched != 1 || topFrame[0].m_pFrame == null)
+            {
+                return logMessage;
+            }
+
+            HashSet<string> tokensUsed = new HashSet<string>();
+            string interpolatedLogMessage = regex.Replace(logMessage, (match) =>
+            {
+                if (match.Success && !string.IsNullOrWhiteSpace(match.Value))
+                {
+                    char c = match.Value[0];
+                    if (c == '$')
+                    {
+                        string token = match.Value.Substring(1);
+                        switch (token)
+                        {
+                            case "FILEPOS":
+                            case "FUNCTION":
+                            case "ADDRESS":
+                            case "TID":
+                            case "TNAME":
+                            case "PID":
+                            case "PNAME":
+                            case "CALLER":
+                            case "CALLSTACK":
```

Comment:
Personally I would implement these - they are the main reason I use tracepoints. Even more than evaluating expressions.

Reply:
In case it isn't obvious what I was pointing to, I meant `$CALLSTACK`.